### PR TITLE
curl: 8.5.0 -> 8.6.0, build with libpsl

### PR DESCRIPTION
**Pull request by mweinelt**
## Description of changes

Updates curl to 8.6.0
- https://curl.se/changes.html#8_6_0
- https://curl.se/docs/CVE-2024-0853.html

Enables libpsl on `curl`.
- https://daniel.haxx.se/blog/2024/01/10/psl-in-curl/

Disables testing libpsl with valgrind, easiest way to resolve cyclic dependencies.

Status: Running build tests

## Things done

- Built on platform(s)
  - [x] x86_64-linux up to `curlFull` and `curl.tests`
  - [x] aarch64-linux up to `curlFull` and `curl.tests`
  - [x] x86_64-darwin up to `curl`
  - [x] aarch64-darwin up to `curlFull` and `curl.tests` w/o `fetchpatch`, which does not eval
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/networking/curl/default.nix [hunk 1]
```diff
@@ -120,6 +118,7 @@ stdenv.mkDerivation (finalAttrs: {
       (lib.withFeature rtmpSupport "librtmp")
       (lib.withFeature rustlsSupport "rustls")
       (lib.withFeature zstdSupport "zstd")
+      (lib.withFeature pslSupport "libpsl")
```
Comment (a-n-n-a-l-e-e) [inline pkgs/tools/networking/curl/default.nix:121]:
```suggestion
```
this gets autodetected when pslSupport is true. (added in 1e4864e29574992c4e51af7eafe8dc65f8181c3d) is there a reason to explicitly enable / disable it?

**Comment by Izorkin** [comment]
I'm getting a build error:
```
  CC       vssh/libcurl_la-libssh.lo
  CC       vssh/libcurl_la-libssh2.lo
  CC       vssh/libcurl_la-wolfssh.lo
  CCLD     libcurl.la
make[2]: Leaving directory '/build/curl-8.6.0/lib'
make[1]: Leaving directory '/build/curl-8.6.0/lib'
Making all in src
make[1]: Entering directory '/build/curl-8.6.0/src'
Making all in ../docs
make[2]: Entering directory '/build/curl-8.6.0/docs'
Making all in .
make[3]: Entering directory '/build/curl-8.6.0/docs'
if test "..x" != "..x" -a -e "./curl.1"; then \
        /nix/store/ddls0pz7c375wlljx6vrrr0px3xvqz3p-coreutils-9.4/bin/install -c -m 644 "./curl.1" /build/curl-8.6.0/docs/curl.1 \
        && touch -r "./curl.1" /build/curl-8.6.0/docs/curl.1; fi
cd cmdline-opts && make
  RENDER  mk-ca-bundle.1
/nix/store/6jli05lcyvb83rwcsr34rn87vrwai8ds-bash-5.2-p21/bin/bash: line 1: ../scripts/cd2nroff: cannot execute: required file not found
make[3]: *** [Makefile:942: mk-ca-bundle.1] Error 127
make[3]: *** Waiting for unfinished jobs....
make[4]: Entering directory '/build/curl-8.6.0/docs/cmdline-opts'
  GENERATE ../../docs/curl.1
make[4]: Leaving directory '/build/curl-8.6.0/docs/cmdline-opts'
make[3]: Leaving directory '/build/curl-8.6.0/docs'
make[2]: *** [Makefile:654: all-recursive] Error 1
make[2]: Leaving directory '/build/curl-8.6.0/docs'
make[1]: *** [Makefile:1553: all-recursive] Error 1
make[1]: Leaving directory '/build/curl-8.6.0/src'
make: *** [Makefile:1262: all-recursive] Error 1
```

File: pkgs/tools/networking/curl/default.nix (hunk 1 above)
Comment (Izorkin) [reply pkgs/tools/networking/curl/default.nix:121]: Curl currently requires explicitly specifying that PSL support should be disabled.
Error without this line:
```
       > checking whether to use builtin CA store of SSL library... yes
       > checking for library containing psl_builtin... no
       > configure: error: libpsl was not found
```

Comment (a-n-n-a-l-e-e) [reply pkgs/tools/networking/curl/default.nix:121]: ah -- i was building 8.5.0.

**Comment by mweinelt** [comment]
What platform? My darwin builds are all still running  (since 4h30m) and no errors.

**Comment by Izorkin** [comment]
> What platform? My darwin builds are all still running (since 4h30m) and no errors.

Fixed, I forgot to add the postPatch block :)

**Comment by mweinelt** [comment]
@ofborg eval

**Comment by adamcstephens** [comment]
This built successfully for me on x86_64-linux and aarch64-darwin. If you have specific packages you'd like me to test let me know, but I wasn't planning on rebuilding everything to run a review on this.

**Comment by mweinelt** [comment]
Check the top post, I built it on all relevant platforms already.

**Review by ghost — APPROVED** [review]

**Comment by risicle** [comment]
Built `curl.tests` (apart from haskell test, too big) and `curlFull.tests.withCheck` on macos 12 x86_64.

**Comment by vcunat** [comment]
Some edge-case doesn't work, apparently and `nixStatic` gets blocked by that:
https://hydra.nixos.org/build/248713644/nixlog/8/tail

**Comment by mweinelt** [comment]
Not entirely sure where this goes wrong.

pkgsStatic.libpsl in `lib/libpsl.a` does have `0000000000000f90 T psl_builtin`.

**Comment by alyssais** [comment]
libpsl depends on libunistring and libidn2.  Static libraries don't embed dependency information, so you have to use some out-of-band mechanism to find their dependencies; usually libtool or pkg-config.  curl is not using pkg-config to find libpsl, so it doesn't pick up the dependencies.

**Comment by alyssais** [comment]
curl does use pkg-config for a bunch of other libraries, so I think it's a case of getting it to do that for libpsl as well.

**Comment by mweinelt** [comment]
I'm talking to upstream, to get the ball rolling on pkg-config support. In the meantime I tried the following, but it fails the same. Not really my area of expertise I must admit.

```patch
diff --git a/pkgs/tools/networking/curl/default.nix b/pkgs/tools/networking/curl/default.nix
index 9cbf6dc1b0fb..89ee10fcc4ad 100644
--- a/pkgs/tools/networking/curl/default.nix
+++ b/pkgs/tools/networking/curl/default.nix
@@ -19,7 +19,7 @@
 , idnSupport ? false, libidn2
 , ldapSupport ? false, openldap
 , opensslSupport ? zlibSupport, openssl
-, pslSupport ? false, libpsl
+, pslSupport ? false, libpsl, libunistring
 , rtmpSupport ? false, rtmpdump
 , scpSupport ? zlibSupport && !stdenv.isSunOS && !stdenv.isCygwin, libssh2
 , wolfsslSupport ? false, wolfssl
@@ -145,6 +145,9 @@ stdenv.mkDerivation (finalAttrs: {
   CXX = "${stdenv.cc.targetPrefix}c++";
   CXXCPP = "${stdenv.cc.targetPrefix}c++ -E";
 
+  NIX_CFLAGS_COMPILE = lib.optionalString (pslSupport && stdenv.hostPlatform.isStatic) "-I${lib.getDev libunistring}/include -I${lib.getDev libidn2}/include";
+  NIX_LDFLAGS= lib.optionalString (pslSupport && stdenv.hostPlatform.isStatic) "-L${lib.getLib libunistring}/lib -L${lib.getLib libidn2}/lib";
+
   # takes 14 minutes on a 24 core and because many other packages depend on curl
   # they cannot be run concurrently and are a bottleneck
   # tests are available in passthru.tests.withCheck
```

**Comment by alyssais** [comment]
The search paths (`-I`, `-L`) should be fine unless the headers are under a subdirectory or something.  What you need to do is force the libraries to be linked: `-lunistring -lidn2`.

Can you link to discussion with upstream?

**Comment by mweinelt** [comment]
The discussion happened on IRC. They are adding pkg-config support to their todo in https://github.com/curl/curl/pull/12919.

**Comment by alyssais** [comment]
Do you have any insight into how likely things from the curl TODO are to end up being done?  Having a NIX_LDFLAGS hack around is not great for maintainability, because dependencies of libraries can change.  I guess for curl we'd probably at least notice via nixStatic though…

**Comment by domenkozar** [comment]
This also brings Python into the curl closure:

❯ grep -R sxr2igfkwhxbagri49b8krmcqz168sim-python3-3.11.8  /nix/store/nwfsn1in46wmgn6ssm2j8famwis5yisz-libpsl-0.21.5
/nix/store/nwfsn1in46wmgn6ssm2j8famwis5yisz-libpsl-0.21.5/bin/psl-make-dafsa:#!/nix/store/sxr2igfkwhxbagri49b8krmcqz168sim-python3-3.11.8/bin/python

**Comment by vcunat** [comment]
Like this?  https://github.com/NixOS/nixpkgs/pull/292260

**Pull request merged**